Add sequence timeout

MissPeppermintPatty opened the pull request:
[why]
As stated by Lars (#3)

> For me, the sequence timeout would be useful mainly for sequences that may be "forgotten" by the operators. For instance, our sequences for joining two XFEL branches may wait for operators to go to the shift leader and ask for some beam shutters to be closed. If the beam shutter never gets closed, the sequence will keep running forever.

[how]
Add new timeout logic to the sequence to abort it when the time elapsed.

Solves issue #3

MissPeppermintPatty commented on tests/test_Sequence.cc:
```diff
@@ -3911,3 +3911,54 @@ TEST_CASE("Sequence: Check setter/getter function for step setup script", "[Sequ
+
+TEST_CASE("Sequence: sequence timeout is lower than step timeout", "[Sequence]")
+{
+    // TODO (Marcus, 13.02.2022):
```
We need more improvement for checking sequence timeout. The current implementation is not enough.

I see two scenarios which would solve the problem as mentioned as a comment in the test case:

1. use another thread when staring a sequence that handles the timeout. But I can imagine that it can be tricky to abort the sequence.
2. use `Context` with two additional fields: one for start time and one for timeout. Both can be checked in the Lua hook `hook_check_timeout_and_termination_request`

But may there are better solutions than the above ones.

MissPeppermintPatty replied:
We go now with the second approach. In `Context` there is a variable of type `TimeoutTrigger` that evaluates a timeout during executing a Lua script. Nevertheless there must also be a sequence timeout check when no Lua script is executed at all.

**My proposal:** check it also in `Sequence::execute_range` (as previously implemented). But I don't know how to implement a test case for that. Any suggestions?

MissPeppermintPatty commented on src/Sequence.cc:
```diff
@@ -415,14 +416,31 @@ Sequence::execute_if_or_elseif_block(Iterator begin, Iterator end, Context& cont
+    using std::chrono::duration;
+
     Iterator step = step_begin;
+    auto start_time = Clock::now();
```
Since we are measuring time differences against a sequence timeout I would encapsulate the functionality into a class instead of using the mutable variables in a sequence or at a Lua hook functionality. The aim is it to have two member functions to control it and hide the mutable part:

```c++
#include "taskolib/Timeout.h"
#include "taskolib/time_types.h"

namespace task {

class TimeoutTrigger {
private:
    Timeout::Duration timeout_{Timeout::infinity()};
    TimePoint start_{Clock::now()};

public:
    void reset(Timeout timeout) { timeout_ = timeout; start_ = Clock::now(); }
    bool is_elapsed() const { return (Clock::now() - start_).count() > timeout_.count(); }
};

} // namespace task
```

What is your meaning about this approach? The code snippet would than change to:

```c++
    ...
    timeout_trigger_.reset(timeout_);

    while(step < step_end)
    {
        if(timeout_trigger_.is_elapsed())
        {
            ...
```

PS: it doesn't have to be with the naming `TimeoutTrigger`. Any other more favourite names is very welcome :-)

alt-graph commented:
I think this looks good overall, but two points come to mind:
- Why is the timeout part of the `Context`? I think it does not logically belong there – I see it rather as an implementation detail of `Sequence`.
- This might be a good occasion to add a "last started" timestamp to the `Sequence`. We already have that for single steps, but IMO it would also be very useful for entire sequences. This might mean that the (nice) `TimeoutTrigger` class becomes superfluous because its single parts (start timestamp and timeout) would exist as single entities in their own right.

MissPeppermintPatty commented:
> Why is the timeout part of the Context? I think it does not logically belong there – I see it rather as an implementation detail of Sequence.

This is a question of design and architecture. The timeout is part of `Context` to minimize changes in the Lua hook function `install_timeout_and_termination_request_hook(.., const Context&, ..)`. I can replace and add two new arguments for timeout and sequence start time but then the function signature becomes more confusing. I know that that is an internal part of the library but nevertheless with eight instead of six arguments it becomes more chaotic. Furthermore with the injection into `Step::execute` it also needs two additional arguments. Therefore I decided to add `TimeoutTrigger` to the `Context` to have the best compromise.

alt-graph commented:
> > Why is the timeout part of the Context? I think it does not logically belong there – I see it rather as an implementation detail of Sequence.
> 
> This is a question of design and architecture. The timeout is part of `Context` to minimize changes in the Lua hook function `install_timeout_and_termination_request_hook(.., const Context&, ..)`. I can replace and add two new arguments for timeout and sequence start time but then the function signature becomes more confusing. I know that that is an internal part of the library but nevertheless with eight instead of six arguments it becomes more chaotic. Furthermore with the injection into `Step::execute` it also needs two additional arguments. Therefore I decided to add `TimeoutTrigger` to the `Context` to have the best compromise.

I still think this needlessly puts implementation details into our public `Context` struct. If we simply put the sequence start time and sequence timeout into `Sequence`, it should be sufficient to hand a `Sequence` pointer to `install_timeout_and_termination_request_hook`:
```cpp
void install_timeout_and_termination_request_hook(sol::state& lua, TimePoint now,
    std::chrono::milliseconds timeout, OptionalStepIndex step_idx,
    const Context& context, const Sequence* sequence, CommChannel* comm_channel)
```
Maybe we could even remove the `timeout` argument, because that might be accessible via the `Step` inside the `Sequence` (although I do not understand all possible consequences right now).

MissPeppermintPatty commented:
> I still think this needlessly puts implementation details into our public `Context` struct. If we simply put the sequence start time and sequence timeout into `Sequence`, it should be sufficient to hand a `Sequence` pointer to `install_timeout_and_termination_request_hook`:
> ```cpp
> void install_timeout_and_termination_request_hook(sol::state& lua, TimePoint now,
>     std::chrono::milliseconds timeout, OptionalStepIndex step_idx,
>     const Context& context, const Sequence* sequence, CommChannel* comm_channel)
> ```
> Maybe we could even remove the `timeout` argument, because that might be accessible via the `Step` inside the `Sequence` (although I do not understand all possible consequences right now).

Hmm, ok, I removed the sequence timeout from `Context`. To solve the problem I added a new function to `lua_details`: `inject_sequence_timeout_to_lua_hook(TimeoutTrigger*)`. It did a proper job on evaluating a sequence timeout if set. When I follow your approach I also have to add additional parameter(s) to `Step::execute()` or add new member function explicitly for the sequence timeout injection. I want to avoid this and we already have lots of parameters in `Step::execute()` with default values. Therefore the class `Step` is still unchanged.

> ...  This might mean that the (nice) TimeoutTrigger class becomes superfluous because its single parts (start timestamp and timeout) would exist as single entities in their own right.

We still have the class `TimeoutTrigger` because I found it much more convenient to gather all data into a class and encapsulate it ('separation of concerns').

Please have a look.

BTW: I also add a new member function at Sequence to get the last executed time: `Sequence::get_time_of_last_execution()`.

alt-graph commented on include/taskolib/Sequence.h:
```diff
@@ -540,18 +566,16 @@ class Sequence
+    bool is_running_{false}; ///< Flag to determine if the sequence is running.
 
+    Timeout timeout_{Timeout::infinity()}; ///< Timeout for sequence.
+    TimeoutTrigger timeout_trigger_;
```
If we want to keep the `TimeoutTrigger` class, I think we should remove the `timeout_` member. It can just as well be stored inside `timeout_trigger_`. This way we just store redundant information and make it unclear which one is valid.

MissPeppermintPatty replied:
Yup! I was also thinking about removing it. Fixed.

alt-graph commented on include/taskolib/TimeoutTrigger.h:
```diff
@@ -0,0 +1,90 @@
+    bool is_elapsed() const
+    {
+        auto timeout_in_ms = static_cast<Timeout::Duration>(timeout_).count();
+        return (Clock::now() - start_).count() > timeout_in_ms;
```
This may be nitpicking, but strictly speaking we should handle the case of infinite timeouts specially. And there is no need for the `count()`:
```
if (!isfinite(timeout_))
    return false;
return Clock::now() - start_ > static_cast<Timeout::Duration>(timeout_);
```

MissPeppermintPatty replied:
You are right. Fixed.

alt-graph commented on src/Sequence.cc:
```diff
@@ -419,6 +425,9 @@ Sequence::Iterator
 {
+    using std::chrono::milliseconds;
+    using std::chrono::duration;
+
```
... `milliseconds` and `duration` are not even used in this function?

MissPeppermintPatty replied:
Residue of adding sequence timeout. Removed.

alt-graph commented on src/lua_details.cc:
```diff
@@ -54,6 +56,14 @@ static const char step_timeout_s_key[] =
 namespace task {
 
+// A holder for a sequence timeout pointer to trigger an abort when the clock elapsed.
+TimeoutTrigger* sequence_timeout_{nullptr};
```
This is a global variable? I guess that won't work at all when running multiple sequences in parallel.

MissPeppermintPatty replied:
Good point! Because if we run multiple sequence consecutively and every sequence sets its own timeout the last one wins. Therefore your approach about adding a `Sequence` pointer to`Step::execute(...)` becomes more or less the only possibility to solve the problem (per default set to `nullptr`). `Step::execute(...)` now forwards the sequence pointer to the Lua hook which is now 'thread-safe'. However, we have a circular dependency of `Sequence` in `Step` which I don't like  ... :-/ Anyhow I add a use case to sequence test that tests the timeout on two parallel sequences during execution.

MissPeppermintPatty replied:
Yup! This resolves the circular dependency :smile:.

Any additional tweaks/enhancements...

alt-graph commented on tests/meson.build:
```diff
@@ -17,6 +17,7 @@ test_src = files(
     'test_Step.cc',
     'test_time_types.cc',
     'test_Timeout.cc',
+    'test_Timeout.cc',
```
?

MissPeppermintPatty replied:
Oops! Nervous finger by copy-and-paste :-). Removed.

alt-graph reviewed (commented):
I like the direction better than before, but I don't think that global variables should be part of the solution.

MissPeppermintPatty commented on tests/test_Sequence.cc:
```diff
@@ -3911,3 +3911,38 @@ TEST_CASE("Sequence: Check setter/getter function for step setup script", "[Sequ
+    auto end = Clock::now();
+
+    // check if executing step (sleeps 10min!) is lower then sequence timeout (500ms+5ms)
+    REQUIRE((end-start).count() < TimePoint(500ms + 5ms).time_since_epoch().count());
```
Ok. Fixed.

alt-graph commented on include/taskolib/Sequence.h:
```diff
@@ -313,6 +319,23 @@ class Sequence
 
+    /**
+     * Returns time of last execution. Will return Epoch time of January 1st 1970 on
+     * a fresh created sequence.
```
Unfortunately the exact date is system-dependent. Maybe we should specify "returns TimePoint{}" for a newly created sequence.

alt-graph commented on include/taskolib/TimeoutTrigger.h:
```diff
@@ -0,0 +1,100 @@
+public:
+    /**
+     * When invoked resets the start time of the internal clock and sets the timeout that
+     * is used in the is_elapsed() member function.
```
Maybe something like "Reset the timeout start time to the current time. The timeout duration remains unchanged." might be a bit easier to understand?

alt-graph commented on include/taskolib/TimeoutTrigger.h:
```diff
@@ -0,0 +1,100 @@
+     * When invoked resets the start time of the internal clock and sets the timeout that
+     * is used in the is_elapsed() member function.
+     *
+     * \param timeout to be measure with the is_elapsed() member function.
```
Which parameter?

alt-graph commented on include/taskolib/TimeoutTrigger.h:
```diff
@@ -0,0 +1,100 @@
+    TimePoint reset() { start_ = Clock::now(); return start_; }
+
+    /**
+     * Get the timeout object.
```
Maybe "duration" instead of "object"?

alt-graph commented on include/taskolib/TimeoutTrigger.h:
```diff
@@ -0,0 +1,100 @@
+     *
+     * \return constant timeout.
+     */
+    const Timeout get_timeout() const { return timeout_; }
```
The first "const" here is meaningless.

alt-graph commented on include/taskolib/TimeoutTrigger.h:
```diff
@@ -0,0 +1,100 @@
+    const Timeout get_timeout() const { return timeout_; }
+
+    /**
+     * Set the timeout object.
```
Also here, "duration" would be more descriptive than "object".

alt-graph commented on include/taskolib/TimeoutTrigger.h:
```diff
@@ -0,0 +1,100 @@
+     *
+     * \return constant start time.
+     */
+    const TimePoint get_start_time() const { return start_; }
```
The first "const" is meaningless.

alt-graph commented on include/taskolib/TimeoutTrigger.h:
```diff
@@ -0,0 +1,100 @@
+    const TimePoint get_start_time() const { return start_; }
+
+    /**
+     * Evaluates if the previous set timeout elapsed.
```
Consider omitting "previous set". Which other timeout could possibly be checked?

alt-graph commented on tests/test_Sequence.cc:
```diff
@@ -3911,3 +3911,77 @@ TEST_CASE("Sequence: Check setter/getter function for step setup script", "[Sequ
+    seq.push_back(std::move(step));
+
+    seq.set_timeout(100ms);
+    REQUIRE(seq.get_timeout() == 100ms);
```
I think the difference between the timeout duration and the sleep duration may be a little too low. Especially on the slow Github CI runners execution can easily be interrupted for 100 ms. So I'd recommend a sleep of some seconds and, to speed the test up, a timeout of 10 ms or so.

alt-graph commented on tests/test_Sequence.cc:
```diff
@@ -3911,3 +3911,77 @@ TEST_CASE("Sequence: Check setter/getter function for step setup script", "[Sequ
+    REQUIRE_THAT(maybe_error->what(), Contains("Timeout: Sequence"));
+}
+
+TEST_CASE("Sequence: sequence timeout is lower than step timeout", "[Sequence]")
```
Maybe we should drop this test case. If I am not mistaken, it does not test anything more than the previous one, except the exact number of milliseconds that have been waited - but unfortunately that can not really be tested because execution times are not deterministic. I'd rather save this 0.5 seconds.

alt-graph commented on tests/test_Sequence.cc:
```diff
@@ -3911,3 +3911,77 @@ TEST_CASE("Sequence: Check setter/getter function for step setup script", "[Sequ
+    gul14::sleep(350ms);
+
+    REQUIRE(seq_1.is_timeout_elapsed() == true);
+    REQUIRE(seq_2.is_timeout_elapsed() == false);
```
This is a race condition. There is no synchronization anywhere to make sure that you can modify `seq_1` and `seq_2` in one thread and simultaneously read them from another thread.

MissPeppermintPatty replied:
Hmm, are you sure? Because both threads execute different sequences without a connection. I thought that a race condition manifest when two threads are executed simultaneously and accessing the same resource.

About the timing issue and as you already mentioned in some other remarks we cannot presume that the execution will possible stop for a while which has an influence on the elapsed time and can fail. 

I will remove the test case but since we have some 'controversy' about the race condition I wanted to keep it to have some remarks from you 😄 . Maybe I overlooked something...

alt-graph commented on tests/test_TimeoutTrigger.cc:
```diff
@@ -0,0 +1,97 @@
+        "TimeoutTrigger is_trivially_destructible");
+
+    TimeoutTrigger timeout_trigger;
+    REQUIRE(timeout_trigger.get_start_time().time_since_epoch().count() == 0L);
```
Just as a heads-up – the following would be equivalent:
```suggestion
    REQUIRE(timeout_trigger.get_start_time() == task::TimePoint{});
```

alt-graph commented on tests/test_TimeoutTrigger.cc:
```diff
@@ -0,0 +1,97 @@
+
+    SECTION("copy assignable")
+    {
+        TimeoutTrigger timeout_trigger_assigned{timeout_trigger};
```
This is copy construction, not copy assignment. Copy assignment would require that both objects have already been constructed:
```cpp
Obj a;
Obj b;
b = a;
```

alt-graph commented on tests/test_TimeoutTrigger.cc:
```diff
@@ -0,0 +1,97 @@
+
+    timeout_trigger.set_timeout(200ms);
+
+    gul14::sleep(100ms);
```
Again, I'd be careful with timing. The best would maybe be to omit the first `sleep()` entirely. We have been bitten by these brittle timing tests over and over again already.

alt-graph reviewed (commented):
Excellent, except for some details in the tests. I'll approve when at least the race condition has been removed/addressed.

alt-graph commented on tests/test_TimeoutTrigger.cc:
```diff
@@ -0,0 +1,105 @@
+/**
+ * \file   test_TriggerTimeout.cc
```
```suggestion
 * \file   test_TimeoutTrigger.cc
```

alt-graph commented on tests/test_TimeoutTrigger.cc:
```diff
@@ -0,0 +1,105 @@
+    SECTION("copy constructable II")
+    {
+        TimeoutTrigger timeout_trigger_copy{timeout_trigger};
+        timeout_trigger_copy.reset();
```
This is by definition the same thing as in lines 55 and 56, so the test is redundant.

alt-graph commented on tests/test_TimeoutTrigger.cc:
```diff
@@ -0,0 +1,105 @@
+
+    timeout_trigger.set_timeout(200ms);
+
+    gul14::sleep(200ms);
```
I would rather make this shorter – the same functionality can be tested with a delay of 10 ms, so why keep ourselves waiting for longer than necessary?

alt-graph reviewed (approved):
Great! Thanks a lot, Marcus! :)
